### PR TITLE
scheduler: fix performance regression at -v3 + contextual logging

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind bug

This was not present in 1.28. The new code added for 1.29 should be fixed before it gets released.

#### What this PR does / why we need it:

The logging instrumentation for contextual logging that was added for 1.29 slowed down the scheduler (i.e. logging verbosity <= 3) by a significant percentage (-28.66% for SchedulingBasic/5000Nodes at -v3) if (and only if!) contextual logging was enabled.

Retrieving the logger from the context causes no measurable slowdown, it's only the various `WithName/WithValues` calls which cause this.

By being more careful about when to use those, the performance impact can be avoided:
- At -v3 or lower, only `WithValues("pod")` is used once per scheduling cycle. This has the intended effect that all log messages for the cycle include the pod information. Once contextual logging is GA, "pod" key/value pairs can be removed from all log calls.
- At -v4 or higher, richer log entries get produced where `WithValues` is also used for the node (when applicable) and `WithName` is used for the current operation and plugin.

#### Special notes for your reviewer:

With these changes, enabling contextual logging causes no measurable slowdown at -v3 or lower. At -v4, the slowdown depends on the test case (-30.51% throughput for SchedulingBasic/5000Nodes, no change for SchedulingCSIPVs/5000Nodes). For some unknown reason (measuring bias?), SchedulingCSIPVs/500Nodes has a ~3& *higher* throughput with contextual logging.

To reproduce benchmarks, use `go test -timeout=0 -count=10  '-bench=BenchmarkPerfScheduling/(SchedulingBasic|SchedulingCSIPVs)'  -benchtime=1ns -run=xxx ./test/integration/scheduler_perf -args -perf-scheduling-label-filter=performance -v=3 -feature-gate=ContextualLogging=false` and ` -feature-gate=ContextualLogging=true`.

#### Does this PR introduce a user-facing change?

```release-note
scheduler: in 1.29 pre-releases, enabling contextual logging slowed down pod scheduling.
```

/cc @mengjiao-liu @alculquicondor @knelasevero @Huang-Wei

### CODE REVIEW

File: pkg/scheduler/framework/runtime/framework.go
Comment (pohly) [inline pkg/scheduler/framework/runtime/framework.go:663]: This is a common pattern in this PR: loop iterations must not change the variable outside of the loop, so local variables get created if (and only if) needed.

When verbose logging is off, this is a cheap copy where the variable then doesn't get modified. When it's on, it gets modified.

**Comment by mengjiao-liu** [comment]
/triage accepted

> With these changes, enabling contextual logging causes no measurable slowdown at -v3 or lower. At -v4, the slowdown depends on the test case (-30.51% throughput for SchedulingBasic/5000Nodes, no change for SchedulingCSIPVs/5000Nodes). For some unknown reason (measuring bias?), SchedulingCSIPVs/500Nodes has a ~3& higher throughput with contextual logging.

What is the environment in which you run the bench test? Are the bench test results basically the same every time?

**Comment by pohly** [comment]
> What is the environment in which you run the bench test? Are the bench test results basically the same every time?

I run on a dedicated Skylake workstation, on Linux, with no UI (i.e. accessed via ssh). I run with `-count=10` and benchstat shows some variation, but mostly the results are stable.

**Review by mengjiao-liu — APPROVED** [review]
LGTM for code implementation.
​
Waiting for confirmation from #sig-scheduling.

**Comment by knelasevero** [comment]
/lgtm

Should we close https://github.com/kubernetes/kubernetes/issues/111672 with this? We still have some duplicated entries that we should remove when this gets out of the gate, no?

**Comment by pohly** [comment]
https://github.com/kubernetes/kubernetes/issues/111672 needs to remain open because we still have duplication.

**Comment by knelasevero** [comment]
Just asking since the TODOs are being removed.

File: pkg/scheduler/framework/runtime/framework.go
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:660]: Overall, the change LGTM, but I wonder if we could do better by delegating to klog.

For example, we could say something like:

```golang
logger = klog.LoggerWithNameAt(logger, "PreFilter", 4)
```

which would just return the same logger if the level is less than 4.

However, I'm not sure what to do about `klog.NewContext`. Should we have a function `klog.ContextAndLoggerWithNameAt`? It sounds ugly, and we would also need a counterpart for `WithLabels` (so 4 new functions in total). Maybe `klog.LoggerWithNameAt` can return a boolean that says whether it created a new logger?

In any case, this discussion is beyond what we can cherry-pick, so we should merge like this anyways. Just food for thought.

File: pkg/scheduler/schedule_one.go
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:148]: this should be even higher in the call stack, in `scheduleOne`.

Comment (pohly) [reply pkg/scheduler/schedule_one.go]: Done: https://github.com/kubernetes/kubernetes/compare/d400a5d743fa38b191c0a9ffd30e6d446573486f..2a23061f6c23b65a916956f2ede2371041944b15

File: pkg/scheduler/framework/runtime/framework.go
Comment (alculquicondor) [reply pkg/scheduler/framework/runtime/framework.go:660]: uhm... maybe it could be a `loggingutils` package within `pkg/scheduler`, where even the constant 4 could be hidden from the developer.

**Review by alculquicondor — COMMENTED** [review]
/approve

**Comment by pohly** [comment]
> Just asking since the TODOs are being removed.

@knelasevero : it's still there, in pkg/scheduler/schedule_one.go.

**Pull request merged**
